solidity/test/maintainer/MaintainerProxy.test.ts

```diff
@@ -0,0 +1,3434 @@
+          authorizedMaintainer.address
+        )
+        const diff = postNotifyThirdPartyBalance.sub(initThirdPartyBalance)
+
+        expect(diff).to.be.gt(0)
+        expect(diff).to.be.lt(
+          ethers.utils.parseUnits("2000000", "gwei") // 0,002 ETH
+        )
+      })
+    })
+  })
+
+  describe("submitDepositSweepProof", () => {
+    before(async () => {
+      await createSnapshot()
+
+      const data: DepositSweepTestData = SingleP2SHDeposit
+      // Take wallet public key hash from first deposit. All
+      // deposits in same sweep batch should have the same value
+      // of that field.
+      const { walletPubKeyHash } = data.deposits[0].reveal
+
+      // Set the deposit dust threshold to 0.0001 BTC, i.e. 100x smaller than
+      // the initial value in the Bridge in order to save test Bitcoins.
+      await bridge.setDepositDustThreshold(10000)
+
+      await bridge.setWallet(walletPubKeyHash, {
+        ecdsaWalletID: ethers.constants.HashZero,
+        mainUtxoHash: ethers.constants.HashZero,
+        pendingRedemptionsValue: 0,
+        createdAt: await lastBlockTime(),
+        movingFundsRequestedAt: 0,
+        closingStartedAt: 0,
+        pendingMovedFundsSweepRequestsCount: 0,
+        state: walletState.Live,
+        movingFundsTargetWalletsCommitmentHash: ethers.constants.HashZero,
+      })
+    })
+
+    after(async () => {
+      await restoreSnapshot()
+    })
+
+    context("when called by an unauthorized third party", async () => {
+      const data: DepositSweepTestData = SingleP2SHDeposit
+
+      before(async () => {
+        await createSnapshot()
+      })
+
+      after(async () => {
+        await restoreSnapshot()
+      })
+
+      it("should revert", async () => {
+        const tx = maintainerProxy
+          .connect(thirdParty)
+          .submitDepositSweepProof(
+            data.sweepTx,
+            data.sweepProof,
+            data.mainUtxo,
+            data.vault
+          )
+
+        await expect(tx).to.be.revertedWith("Caller is not authorized")
+      })
+    })
+
+    context("when called by an authorized maintainer", async () => {
+      before(async () => {
+        await createSnapshot()
+      })
+
+      after(async () => {
+        await restoreSnapshot()
+      })
+
+      context("when there is only one input", () => {
+        context(
+          "when the single input is a revealed unswept P2SH deposit",
+          () => {
+            const data: DepositSweepTestData = SingleP2SHDeposit
+            let sweepOutcome: ScenarioOutcome
+
+            before(async () => {
+              await createSnapshot()
+
+              sweepOutcome = await runDepositSweepScenario(data)
+            })
+
+            after(async () => {
+              await restoreSnapshot()
+            })
+
+            it("should not revert", async () => {
+              await expect(sweepOutcome.tx.wait()).not.to.be.reverted
+            })
```

Comment:
https://github.com/keep-network/tbtc-v2/pull/209/commits/0c1261b569df4efca23555f3408add0172c9c31f

Reply:
Sure, but the main purpose of these tests was to check the refunds. `MaintainerProxy` is essentially a wrapper to the `Bridge` contract. All the details cover other `Bridge` specific tests and I figured it would be enough just to check if a call was executed without any reverts. Problems with the execution of the refunds would be captured by the "should refund ETH" tests too. I also didn't see any try-catch blocks that would call an external contract from the `Bridge`. Anyway, both work for me.

Reply:
The previous approach was not that bad, but the advantage of asserting emitted events is that we also check that the function of `Bridge` was actually called. There are two things `MaintainerProxy` needs to do:
1. Call `Bridge` function.
2. Refund ether.

Asserting only that the transaction does not revert may not check properly point 1 if, for example, gas offset is too low or the balance change assertion is not precise enough.